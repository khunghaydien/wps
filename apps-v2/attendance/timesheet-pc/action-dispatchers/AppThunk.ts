import { Action } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';

import { State } from '../modules';

export type AppAction<ReturnType = unknown> = ThunkAction<
  ReturnType,
  State,
  void,
  Action<unknown>
>;

export type AppDispatch = ThunkDispatch<
  State,
  void,
  Action<unknown | AppAction>
>;
