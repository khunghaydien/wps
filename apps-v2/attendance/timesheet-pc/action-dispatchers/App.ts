import isNil from 'lodash/isNil';

import {
  catchApiError,
  catchBusinessError,
  withLoading,
} from '@apps/commons/actions/app';
import { getUserSetting } from '@apps/commons/actions/userSetting';
import msg from '@apps/commons/languages';
import { setUserPermission } from '@apps/commons/modules/accessControl/permission';
import { actions as proxyEmployeeInfoActions } from '@apps/commons/modules/proxyEmployeeInfo';
import { actions as standaloneModeActions } from '@apps/commons/modules/standaloneMode';
import UrlUtil from '@apps/commons/utils/UrlUtil';

import EmployeeRepository from '@apps/repositories/EmployeeRepository';

import { Permission } from '@apps/domain/models/access-control/Permission';
import { UserSetting } from '@apps/domain/models/UserSetting';

import { initDailyStampTime } from '@attendance/timesheet-pc/action-dispatchers/StampWidget';

import UseCases from '../UseCases';
import { fetchTimesheetAndAvailableDailyRequest } from './Timesheet';
import { IOutputData as IFetchTimesheetOutputData } from '@attendance/domain/useCases/timesheet/IFetchUseCase';
import Events from '@attendance/timesheet-pc/events';

/**
 * 勤務表画面を初期化する
 */
export const initialize =
  (param: { userPermission: Permission }) =>
  async (dispatch): Promise<void> => {
    const urlQuery = UrlUtil.getUrlQuery();
    const empId = urlQuery && urlQuery.empId;
    const targetDate = urlQuery && urlQuery.targetDate;
    const standalone = (urlQuery && urlQuery.standalone) === '1';
    const isProxy = !isNil(empId);
    if (isProxy && isNil(targetDate)) {
      dispatch(
        catchBusinessError(
          msg().Com_Err_ErrorTitle,
          msg().Com_Err_ProxyPermissionErrorBody,
          msg().Com_Err_ProxyPermissionErrorSolution,
          { isContinuable: false }
        )
      );
    }
    if (standalone) {
      dispatch(standaloneModeActions.enable());
    }

    if (param) {
      dispatch(setUserPermission(param.userPermission));
    }

    await dispatch(
      withLoading(async (): Promise<void> => {
        try {
          // @ts-ignore
          // Cast (void | UserSetting) to UserSetting
          const [userSetting, fetchedTimesheet, _]: [
            UserSetting,
            IFetchTimesheetOutputData,
            void
          ] = await Promise.all([
            dispatch(getUserSetting()),
            UseCases().fetchTimesheet({
              targetDate,
              employeeId: empId,
            }),
            dispatch(async () => {
              if (!isProxy) {
                await dispatch(initDailyStampTime());
              }
            }),
          ]);

          const { timesheet } = fetchedTimesheet;

          const isSuccess = !isNil(timesheet);

          if (!isSuccess) {
            return;
          }

          if (timesheet.isMigratedSummary) {
            return;
          }

          const isCurrentEmployee =
            isNil(empId) || userSetting.employeeId === empId;

          if (!isCurrentEmployee) {
            const [targetEmployee] = await EmployeeRepository.search({
              id: empId || undefined,
              targetDate: targetDate || undefined,
            });
            dispatch(
              proxyEmployeeInfoActions.set({
                id: targetEmployee.id,
                employeeCode: targetEmployee.code,
                employeeName: targetEmployee.name,
                employeePhotoUrl: targetEmployee.user.photoUrl,
                departmentCode: '', // Not used?
                departmentName: targetEmployee.department
                  ? targetEmployee.department.name
                  : '',
                title: targetEmployee.title,
                managerName: targetEmployee.manager
                  ? targetEmployee.manager.name
                  : '',
              })
            );
          }
          Events.initialized.publish({
            fetchedUserSetting: userSetting,
            fetchedTimesheet,
          });
        } catch (e) {
          dispatch(catchApiError(e, { isContinuable: false }));
        }
      })
    );
  };

/**
 * @param {String} targetDate the target date
 * @param {Object} targetEmployee the target employee
 */
export const switchProxyEmployee =
  (targetDate: string, targetEmployee: { id: string }) =>
  async (dispatch): Promise<boolean> => {
    try {
      await dispatch(
        withLoading(async (): Promise<void> => {
          await UseCases().fetchTimesheet({
            targetDate,
            employeeId: targetEmployee.id,
          });
          dispatch(proxyEmployeeInfoActions.set(targetEmployee));
        })
      );
      return true;
    } catch (e) {
      if (e && e.errorCode === 'ATT_DAILY_IMPORT_REGISTERING') {
        dispatch(
          catchBusinessError(
            msg().Com_Lbl_Error,
            msg().Att_Msg_DailyImportProcessing,
            null,
            { isContinuable: true }
          )
        );
      } else {
        dispatch(
          catchBusinessError(
            msg().Com_Lbl_Error,
            msg().Com_Err_ProxyPermissionErrorBody,
            msg().Com_Err_ProxyPermissionErrorSolution,
            { isContinuable: true }
          )
        );
        return false;
      }
    }
  };

/**
 * 承認者を変更した後に実行する
 * 将来的にEventService.publish() に置換する。
 */
export const onChangeApproverEmployee =
  (
    targetDate: string | null | undefined = null,
    endDate: string | null | undefined = null
  ) =>
  async (dispatch) =>
    dispatch(
      withLoading(async () => {
        await dispatch(
          fetchTimesheetAndAvailableDailyRequest(targetDate, endDate)
        );
      })
    );

/**
 * 打刻した後に呼び出す一連のアクション
 */
export const onStampSuccess = () => (_) => {
  Events.stampedTime.publish();
};

export default initialize;
