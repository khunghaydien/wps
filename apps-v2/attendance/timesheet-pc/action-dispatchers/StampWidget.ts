import * as dailyStampTimeResultActions from '../../../commons/action-dispatchers/DailyStampTimeResult';
import * as commonActions from '../../../commons/actions/app';
import { AppDispatch } from '../../../commons/modules/AppThunk';

import DailyStampTimeRepository from '@attendance/repositories/DailyStampTimeRepository';

import {
  CLOCK_TYPE,
  EditingDailyStampTime,
  MODE_TYPE,
  ModeType,
  STAMP_SOURCE,
} from '@attendance/domain/models/DailyStampTime';

import { actions as stampWidgetActions } from '../modules/entities/stampWidget';

const MODE_TYPE_MAP = {
  [MODE_TYPE.CLOCK_IN]: CLOCK_TYPE.IN,
  [MODE_TYPE.CLOCK_OUT]: CLOCK_TYPE.OUT,
  [MODE_TYPE.CLOCK_REIN]: CLOCK_TYPE.REIN,
};

/**
 * 打刻情報を取得する
 */
export const initDailyStampTime = () => (dispatch: AppDispatch) => {
  return DailyStampTimeRepository.fetch()
    .then((result) => {
      // 打刻情報を適用する
      dispatch(stampWidgetActions.applyDailyStampTime(result));
    })
    .catch((err) =>
      dispatch(commonActions.catchApiError(err, { isContinuable: true }))
    );
};

/**
 * 打刻を実行（送信）する
 */
export const submitStamp =
  (
    stampWidget: EditingDailyStampTime,
    options:
      | {
          withGlobalLoading?: boolean;
          onStampSuccess?: () => void;
        }
      | null
      | undefined
  ) =>
  (dispatch: AppDispatch) => {
    // NOTE: 本来発生しえない／mode未定だと打刻ボタンが非活性なので
    if (stampWidget.mode === null || stampWidget.mode === undefined) {
      // eslint-disable-next-line prefer-promise-reject-errors
      return Promise.reject();
    }

    const { withGlobalLoading, onStampSuccess } = options || {};

    const commonPostProcess = () =>
      DailyStampTimeRepository.post({
        clockType: MODE_TYPE_MAP[stampWidget.mode],
        comment: stampWidget.message,
        source: STAMP_SOURCE.WEB,
      });
    const postProcess = withGlobalLoading
      ? commonActions.withLoading(commonPostProcess)
      : commonPostProcess;

    // ユーザー操作を防止する
    dispatch(stampWidgetActions.blockOperation());

    return dispatch(postProcess)
      .then((result) =>
        dispatch(
          dailyStampTimeResultActions.confirmToComplementInsufficientingRestTime(
            result
          )
        )
      )
      .then(() => {
        if (onStampSuccess && onStampSuccess instanceof Function) {
          onStampSuccess();
        }
      })
      .catch((err) =>
        dispatch(commonActions.catchApiError(err, { isContinuable: true }))
      )
      .then(() => {
        dispatch(initDailyStampTime());
      });
  };

/**
 * 出勤・退勤を切り替える
 * @param {String} clockType
 */
export const switchModeType =
  (clockType: ModeType) => (dispatch: AppDispatch) => {
    dispatch(stampWidgetActions.switchModeType(clockType));
  };

/**
 * メッセージを更新する
 * @param {String} message
 * @returns {{type: String, payload: Object}}
 */
export const updateMessage = (message: string) => (dispatch: AppDispatch) => {
  dispatch(stampWidgetActions.updateMessage(message));
};
