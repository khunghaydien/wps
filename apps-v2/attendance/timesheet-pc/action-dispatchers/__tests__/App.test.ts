import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { getUserSetting } from '@apps/commons/actions/userSetting';
import { actions as standaloneModeActions } from '@apps/commons/modules/standaloneMode';

import EmployeeRepository from '@apps/repositories/EmployeeRepository';

import { Permission } from '@apps/domain/models/access-control/Permission';

import { AppDispatch } from '@attendance/timesheet-pc/action-dispatchers/AppThunk';
import { initDailyStampTime } from '@attendance/timesheet-pc/action-dispatchers/StampWidget';

import UseCases from '../../UseCases';
import * as App from '../App';
import EventEmitter from '@apps/attendance/libraries/Event/emitter';
import Events from '@attendance/timesheet-pc/events';

const mockStore = configureMockStore([thunk]);

jest.mock('../Timesheet');
jest.mock('../../../../commons/actions/userSetting');
jest.mock('../StampWidget');
jest.mock('../TimeTrackAlert');
jest.mock('../DailyTimeTrack');
jest.mock('../../../../repositories/EmployeeRepository');
jest.mock('../../UseCases');
jest.mock('@attendance/libraries/Event/emitter');

const spyOnStandaloneModeActionsEnable = jest.spyOn(
  standaloneModeActions,
  'enable'
);

const reloadTimesheet = () => ({
  type: 'RELOAD_TIMESHEET',
});

beforeEach(() => {
  jest.clearAllMocks();
  (UseCases().fetchTimesheet as unknown as jest.Mock).mockReset();
  Object.defineProperty(window, 'location', {
    get: () => ({
      search: '',
    }),
  });
  (getUserSetting as jest.Mock).mockImplementation(
    (...args) =>
      (dispatch: AppDispatch) => {
        dispatch({
          type: 'GET_USER_SETTING',
          payload: args,
        });
        return Promise.resolve({});
      }
  );
  (initDailyStampTime as jest.Mock).mockImplementation(
    (...args) =>
      (dispatch: AppDispatch) => {
        dispatch({ type: 'INIT_DAILY_STAMP_TIME', payload: args });
        return Promise.resolve();
      }
  );
});

describe('initialize()', () => {
  it('should do with Login User.', async () => {
    // Arrange
    const store = mockStore({});
    const dispatch = store.dispatch as AppDispatch;
    const fetchedTimesheet = {
      timesheet: {
        workingTypeList: [
          {
            useAllowanceManagement: false,
          },
        ],
      },
    };
    (UseCases().fetchTimesheet as unknown as jest.Mock).mockImplementationOnce(
      () => {
        dispatch(reloadTimesheet());
        return Promise.resolve(fetchedTimesheet);
      }
    );

    // Act
    await dispatch(App.initialize({ userPermission: {} as Permission }));

    // Assert
    expect(store.getActions()).toMatchSnapshot();
    expect(spyOnStandaloneModeActionsEnable).toBeCalledTimes(0);
    expect(UseCases().fetchTimesheet).toBeCalledWith({
      targetDate: null,
      employeeId: null,
    });
    expect(initDailyStampTime).toBeCalledTimes(1);
    expect(EmployeeRepository.search).toBeCalledTimes(0);
    expect(EventEmitter.publish).toBeCalledWith(Events.initialized.eventName, {
      fetchedUserSetting: {},
      fetchedTimesheet,
    });
  });

  it('should return if timesheet is null.', async () => {
    // Arrange
    const store = mockStore({});
    const dispatch = store.dispatch as AppDispatch;
    const fetchedTimesheet = {
      timesheet: null,
    };
    (UseCases().fetchTimesheet as unknown as jest.Mock).mockImplementationOnce(
      () => {
        dispatch(reloadTimesheet());
        return Promise.resolve(fetchedTimesheet);
      }
    );

    // Act
    await dispatch(App.initialize({ userPermission: {} as Permission }));

    // Assert
    expect(store.getActions()).toMatchSnapshot();
    expect(EventEmitter.publish).not.toBeCalledWith(
      Events.initialized.eventName,
      {
        fetchedUserSetting: {},
        fetchedTimesheet,
      }
    );
  });

  it('should return if isMigratedSummary is false.', async () => {
    // Arrange
    const store = mockStore({});
    const dispatch = store.dispatch as AppDispatch;
    const fetchedTimesheet = {
      timesheet: {
        isMigratedSummary: true,
        startDate: '2020-01-01',
        endDate: '2020-01-31',
      },
    };
    (UseCases().fetchTimesheet as unknown as jest.Mock).mockImplementationOnce(
      () => {
        dispatch(reloadTimesheet());
        return Promise.resolve(fetchedTimesheet);
      }
    );

    // Act
    await dispatch(App.initialize({ userPermission: {} as Permission }));

    // Assert
    expect(store.getActions()).toMatchSnapshot();
    expect(EventEmitter.publish).not.toBeCalledWith(
      Events.initialized.eventName,
      {
        fetchedUserSetting: {},
        fetchedTimesheet,
      }
    );
  });

  it('should return error if UseCases().reloadTimesheet() throw Error.', async () => {
    // Arrange
    const store = mockStore({});
    const dispatch = store.dispatch as AppDispatch;
    const error = new Error('Test Error');
    error.stack = null;
    (UseCases().fetchTimesheet as unknown as jest.Mock).mockImplementationOnce(
      () => {
        dispatch(reloadTimesheet());
        return Promise.reject(error);
      }
    );
    // Act
    await dispatch(App.initialize({ userPermission: {} as Permission }));

    // Assert
    expect(store.getActions()).toMatchSnapshot();
    expect(EventEmitter.publish).not.toHaveBeenCalled();
  });

  it('should do with Login User and specified date.', async () => {
    // Arrange
    const store = mockStore({});
    const dispatch = store.dispatch as AppDispatch;
    const targetDate = '2021-12-01';
    const fetchedTimesheet = {
      timesheet: {
        workingTypeList: [
          {
            useAllowanceManagement: false,
          },
        ],
      },
    };

    Object.defineProperty(window, 'location', {
      get: () => ({
        search: `?targetDate=${targetDate}`,
      }),
    });
    (UseCases().fetchTimesheet as unknown as jest.Mock).mockImplementationOnce(
      () => {
        dispatch(reloadTimesheet());
        return Promise.resolve(fetchedTimesheet);
      }
    );

    // Act
    await dispatch(App.initialize({ userPermission: {} as Permission }));

    // Assert
    expect(store.getActions()).toMatchSnapshot();
    expect(spyOnStandaloneModeActionsEnable).toBeCalledTimes(0);
    expect(UseCases().fetchTimesheet).toBeCalledWith({
      targetDate,
      employeeId: undefined,
    });
    expect(initDailyStampTime).toBeCalledTimes(1);
    expect(EmployeeRepository.search).toBeCalledTimes(0);
    expect(EventEmitter.publish).toBeCalledWith(Events.initialized.eventName, {
      fetchedUserSetting: {},
      fetchedTimesheet,
    });
  });

  it('should not do if error is occurred.', async () => {
    // Arrange
    const store = mockStore({});
    const dispatch = store.dispatch as AppDispatch;
    (getUserSetting as jest.Mock).mockImplementationOnce(() => () => {
      const error = new Error();
      error.stack = null;
      throw error;
    });
    (UseCases().fetchTimesheet as unknown as jest.Mock).mockImplementationOnce(
      () => {
        dispatch(reloadTimesheet());
        return Promise.resolve({
          timesheet: {
            workingTypeList: [
              {
                useAllowanceManagement: false,
              },
            ],
          },
        });
      }
    );

    // Act
    await dispatch(App.initialize({ userPermission: {} as Permission }));

    // Assert
    expect(store.getActions()).toMatchSnapshot();
    expect(EventEmitter.publish).not.toHaveBeenCalled();
  });

  describe('with parameters', () => {
    it('should do with proxy.', async () => {
      // Arrange
      const store = mockStore({});
      const dispatch = store.dispatch as AppDispatch;
      const employeeId = 'empId';
      const targetDate = '2021-12-01';
      Object.defineProperty(window, 'location', {
        get: () => ({
          search: `?empId=${employeeId}&targetDate=${targetDate}`,
        }),
      });
      const fetchedUserSetting = {
        useWorkTime: true,
      };
      (getUserSetting as jest.Mock).mockImplementationOnce(
        (...args) =>
          (dispatch: AppDispatch) => {
            // @ts-ignore
            dispatch(getUserSetting(...args));
            return fetchedUserSetting;
          }
      );
      const fetchedTimesheet = {
        timesheet: {
          workingTypeList: [
            {
              useAllowanceManagement: false,
            },
          ],
        },
      };
      (
        UseCases().fetchTimesheet as unknown as jest.Mock
      ).mockImplementationOnce(() => {
        dispatch(reloadTimesheet());
        return Promise.resolve(fetchedTimesheet);
      });

      // Act
      await dispatch(
        App.initialize({
          userPermission: {
            viewTimeTrackByDelegate: true,
          } as Permission,
        })
      );

      // Assert
      expect(store.getActions()).toMatchSnapshot();
      expect(spyOnStandaloneModeActionsEnable).toBeCalledTimes(0);
      expect(UseCases().fetchTimesheet).toBeCalledWith({
        targetDate,
        employeeId,
      });
      expect(initDailyStampTime).toBeCalledTimes(0);
      expect(EmployeeRepository.search).toBeCalledTimes(1);
      expect(EmployeeRepository.search).toBeCalledWith({
        id: employeeId,
        targetDate,
      });
      expect(EventEmitter.publish).toBeCalledWith(
        Events.initialized.eventName,
        {
          fetchedUserSetting,
          fetchedTimesheet,
        }
      );
    });

    it('should do with proxy and standalone.', async () => {
      // Arrange
      const store = mockStore({});
      const dispatch = store.dispatch as AppDispatch;
      const employeeId = 'empId';
      const targetDate = '2021-12-01';
      const standalone = '1';
      Object.defineProperty(window, 'location', {
        get: () => ({
          search: `?empId=${employeeId}&targetDate=${targetDate}&standalone=${standalone}`,
        }),
      });
      const fetchedUserSetting = {
        useWorkTime: true,
      };
      (getUserSetting as jest.Mock).mockImplementationOnce(
        (...args) =>
          (dispatch: AppDispatch) => {
            // @ts-ignore
            dispatch(getUserSetting(...args));
            return fetchedUserSetting;
          }
      );
      const fetchedTimesheet = {
        timesheet: {
          workingTypeList: [
            {
              useAllowanceManagement: false,
            },
          ],
        },
      };
      (
        UseCases().fetchTimesheet as unknown as jest.Mock
      ).mockImplementationOnce(() => {
        dispatch(reloadTimesheet());
        return Promise.resolve(fetchedTimesheet);
      });

      // Act
      await dispatch(
        App.initialize({
          userPermission: {
            viewTimeTrackByDelegate: true,
          } as Permission,
        })
      );

      // Assert
      expect(store.getActions()).toMatchSnapshot();
      expect(spyOnStandaloneModeActionsEnable).toBeCalledTimes(1);
      expect(spyOnStandaloneModeActionsEnable).toBeCalledWith();
      expect(UseCases().fetchTimesheet).toBeCalledWith({
        targetDate,
        employeeId,
      });
      expect(initDailyStampTime).toBeCalledTimes(0);
      expect(EmployeeRepository.search).toBeCalledTimes(1);
      expect(EmployeeRepository.search).toBeCalledWith({
        id: employeeId,
        targetDate,
      });
      expect(EventEmitter.publish).toBeCalledWith(
        Events.initialized.eventName,
        {
          fetchedUserSetting,
          fetchedTimesheet,
        }
      );
    });

    it('should not do if targetDate is null with proxy.', async () => {
      // Arrange
      const store = mockStore({});
      const dispatch = store.dispatch as AppDispatch;
      const employeeId = 'empId';
      Object.defineProperty(window, 'location', {
        get: () => ({
          search: `?empId=${employeeId}`,
        }),
      });
      const fetchedUserSetting = {
        useWorkTime: true,
      };
      (getUserSetting as jest.Mock).mockImplementationOnce(
        (...args) =>
          (dispatch: AppDispatch) => {
            // @ts-ignore
            dispatch(getUserSetting(...args));
            return fetchedUserSetting;
          }
      );
      const fetchedTimesheet = {
        timesheet: {
          workingTypeList: [
            {
              useAllowanceManagement: false,
            },
          ],
        },
      };
      (
        UseCases().fetchTimesheet as unknown as jest.Mock
      ).mockImplementationOnce(() => {
        dispatch(reloadTimesheet());
        return Promise.resolve(fetchedTimesheet);
      });

      // Act
      await dispatch(
        App.initialize({
          userPermission: {
            viewTimeTrackByDelegate: true,
          } as Permission,
        })
      );

      // Assert
      expect(store.getActions()).toMatchSnapshot();
      expect(spyOnStandaloneModeActionsEnable).toBeCalledTimes(0);
      expect(UseCases().fetchTimesheet).toBeCalledWith({
        targetDate: undefined,
        employeeId,
      });
      expect(initDailyStampTime).toBeCalledTimes(0);
      expect(EmployeeRepository.search).toBeCalledTimes(1);
      expect(EmployeeRepository.search).toBeCalledWith({
        id: employeeId,
        targetDate: undefined,
      });
      expect(EventEmitter.publish).toBeCalledWith(
        Events.initialized.eventName,
        {
          fetchedUserSetting,
          fetchedTimesheet,
        }
      );
    });
  });
});

describe('switchProxyEmployee()', () => {
  it('should do.', async () => {
    // Arrange
    const store = mockStore({});
    const dispatch = store.dispatch as AppDispatch;
    const targetDate = '2022-01-01';
    const employeeId = 'employeeId';
    (UseCases().fetchTimesheet as unknown as jest.Mock).mockImplementationOnce(
      () => {
        dispatch(reloadTimesheet());
        return Promise.resolve({
          timesheet: {
            workingTypeList: [
              {
                useAllowanceManagement: false,
              },
            ],
          },
        });
      }
    );

    // Act
    const result = await dispatch(
      App.switchProxyEmployee(targetDate, { id: employeeId })
    );

    // Assert
    expect(result).toBe(true);
    expect(store.getActions()).toMatchSnapshot();
    expect(UseCases().fetchTimesheet).toBeCalledTimes(1);
    expect(UseCases().fetchTimesheet).toBeCalledWith({
      targetDate,
      employeeId,
    });
  });

  it('should return if isMigratedSummary is false.', async () => {
    // Arrange
    const store = mockStore({});
    const dispatch = store.dispatch as AppDispatch;
    (UseCases().fetchTimesheet as unknown as jest.Mock).mockImplementationOnce(
      () => {
        dispatch(reloadTimesheet());
        return Promise.resolve({
          timesheet: {
            isMigratedSummary: true,
            startDate: '2020-01-01',
            endDate: '2020-01-31',
          },
        });
      }
    );

    // Act
    await dispatch(App.onStampSuccess());

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });

  it('should not do if UseCases().reloadTimesheet throw Error.', async () => {
    // Arrange
    const store = mockStore({});
    const dispatch = store.dispatch as AppDispatch;
    const targetDate = '2022-01-01';
    const employeeId = 'employeeId';
    const error = new Error('Test Error');
    error.stack = null;
    (UseCases().fetchTimesheet as unknown as jest.Mock).mockImplementationOnce(
      () => {
        dispatch(reloadTimesheet());
        return Promise.reject(error);
      }
    );

    // Act
    const result = await dispatch(
      App.switchProxyEmployee(targetDate, { id: employeeId })
    );

    // Assert
    expect(result).toBe(false);
    expect(store.getActions()).toMatchSnapshot();
  });

  it('should not do if error is occurred.', async () => {
    // Arrange
    const store = mockStore({});
    const dispatch = store.dispatch as AppDispatch;
    const targetDate = '2022-01-01';
    const employeeId = 'employeeId';
    (UseCases().fetchTimesheet as unknown as jest.Mock).mockImplementationOnce(
      () => {
        dispatch(reloadTimesheet());
        const error = new Error();
        error.stack = null;
        throw error;
      }
    );

    // Act
    const result = await dispatch(
      App.switchProxyEmployee(targetDate, { id: employeeId })
    );

    // Assert
    expect(result).toBe(false);
    expect(store.getActions()).toMatchSnapshot();
  });
});

describe('onStampSuccess()', () => {
  it('should do.', async () => {
    // Arrange
    const store = mockStore({});
    const dispatch = store.dispatch as AppDispatch;

    // Act
    await dispatch(App.onStampSuccess());

    // Assert
    expect(store.getActions()).toMatchSnapshot();
    expect(EventEmitter.publish).toBeCalledTimes(1);
    expect(EventEmitter.publish).toBeCalledWith(
      Events.stampedTime.eventName,
      undefined
    );
  });
});
