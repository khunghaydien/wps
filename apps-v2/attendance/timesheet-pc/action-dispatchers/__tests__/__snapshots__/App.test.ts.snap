// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initialize() should do with Login User and specified date. 1`] = `
Array [
  Object {
    "payload": Object {},
    "type": "COMMONS/ACCESS_CONTROL/SET_USER_PERMISSION",
  },
  Object {
    "type": "LOADING_START",
  },
  Object {
    "payload": Array [],
    "type": "GET_USER_SETTING",
  },
  Object {
    "type": "RELOAD_TIMESHEET",
  },
  Object {
    "payload": Array [],
    "type": "INIT_DAILY_STAMP_TIME",
  },
  Object {
    "type": "LOADING_END",
  },
]
`;

exports[`initialize() should do with Login User. 1`] = `
Array [
  Object {
    "payload": Object {},
    "type": "COMMONS/ACCESS_CONTROL/SET_USER_PERMISSION",
  },
  Object {
    "type": "LOADING_START",
  },
  Object {
    "payload": Array [],
    "type": "GET_USER_SETTING",
  },
  Object {
    "type": "RELOAD_TIMESHEET",
  },
  Object {
    "payload": Array [],
    "type": "INIT_DAILY_STAMP_TIME",
  },
  Object {
    "type": "LOADING_END",
  },
]
`;

exports[`initialize() should not do if error is occurred. 1`] = `
Array [
  Object {
    "payload": Object {},
    "type": "COMMONS/ACCESS_CONTROL/SET_USER_PERMISSION",
  },
  Object {
    "type": "LOADING_START",
  },
  Object {
    "payload": FatalError {
      "errorCode": undefined,
      "message": "",
      "name": "Error",
      "stacktrace": null,
      "type": "FatalError",
    },
    "type": "CATCH_UNEXPECTED_ERROR",
  },
  Object {
    "type": "LOADING_END",
  },
]
`;

exports[`initialize() should return error if UseCases().reloadTimesheet() throw Error. 1`] = `
Array [
  Object {
    "payload": Object {},
    "type": "COMMONS/ACCESS_CONTROL/SET_USER_PERMISSION",
  },
  Object {
    "type": "LOADING_START",
  },
  Object {
    "payload": Array [],
    "type": "GET_USER_SETTING",
  },
  Object {
    "type": "RELOAD_TIMESHEET",
  },
  Object {
    "payload": Array [],
    "type": "INIT_DAILY_STAMP_TIME",
  },
  Object {
    "payload": FatalError {
      "errorCode": undefined,
      "message": "Test Error",
      "name": "Error",
      "stacktrace": null,
      "type": "FatalError",
    },
    "type": "CATCH_UNEXPECTED_ERROR",
  },
  Object {
    "type": "LOADING_END",
  },
]
`;

exports[`initialize() should return if isMigratedSummary is false. 1`] = `
Array [
  Object {
    "payload": Object {},
    "type": "COMMONS/ACCESS_CONTROL/SET_USER_PERMISSION",
  },
  Object {
    "type": "LOADING_START",
  },
  Object {
    "payload": Array [],
    "type": "GET_USER_SETTING",
  },
  Object {
    "type": "RELOAD_TIMESHEET",
  },
  Object {
    "payload": Array [],
    "type": "INIT_DAILY_STAMP_TIME",
  },
  Object {
    "type": "LOADING_END",
  },
]
`;

exports[`initialize() should return if timesheet is null. 1`] = `
Array [
  Object {
    "payload": Object {},
    "type": "COMMONS/ACCESS_CONTROL/SET_USER_PERMISSION",
  },
  Object {
    "type": "LOADING_START",
  },
  Object {
    "payload": Array [],
    "type": "GET_USER_SETTING",
  },
  Object {
    "type": "RELOAD_TIMESHEET",
  },
  Object {
    "payload": Array [],
    "type": "INIT_DAILY_STAMP_TIME",
  },
  Object {
    "type": "LOADING_END",
  },
]
`;

exports[`initialize() with parameters should do with proxy and standalone. 1`] = `
Array [
  Object {
    "type": "COMMONS/MODULES/STANDALONEMODE/ENABLE",
  },
  Object {
    "payload": Object {
      "viewTimeTrackByDelegate": true,
    },
    "type": "COMMONS/ACCESS_CONTROL/SET_USER_PERMISSION",
  },
  Object {
    "type": "LOADING_START",
  },
  Object {
    "payload": Array [],
    "type": "GET_USER_SETTING",
  },
  Object {
    "type": "RELOAD_TIMESHEET",
  },
  Object {
    "payload": Object {
      "departmentCode": "",
      "departmentName": "Department Name",
      "employeeCode": "CODE",
      "employeeName": "Employee Name",
      "employeePhotoUrl": "photoUrl",
      "id": "ID",
      "managerName": "Manager Name",
      "title": "Title",
    },
    "type": "COMMONS/MODULES/PROXY_EMP_INFO/SET",
  },
  Object {
    "type": "LOADING_END",
  },
]
`;

exports[`initialize() with parameters should do with proxy. 1`] = `
Array [
  Object {
    "payload": Object {
      "viewTimeTrackByDelegate": true,
    },
    "type": "COMMONS/ACCESS_CONTROL/SET_USER_PERMISSION",
  },
  Object {
    "type": "LOADING_START",
  },
  Object {
    "payload": Array [],
    "type": "GET_USER_SETTING",
  },
  Object {
    "type": "RELOAD_TIMESHEET",
  },
  Object {
    "payload": Object {
      "departmentCode": "",
      "departmentName": "Department Name",
      "employeeCode": "CODE",
      "employeeName": "Employee Name",
      "employeePhotoUrl": "photoUrl",
      "id": "ID",
      "managerName": "Manager Name",
      "title": "Title",
    },
    "type": "COMMONS/MODULES/PROXY_EMP_INFO/SET",
  },
  Object {
    "type": "LOADING_END",
  },
]
`;

exports[`initialize() with parameters should not do if targetDate is null with proxy. 1`] = `
Array [
  Object {
    "payload": [Error: Error - The timetable could not be obtained. There is a possibility that the target employee does not exist on the start date of this month, or the authority necessary to switch to the target employee is insufficient.],
    "type": "CATCH_BUSINESS_ERROR",
  },
  Object {
    "payload": Object {
      "viewTimeTrackByDelegate": true,
    },
    "type": "COMMONS/ACCESS_CONTROL/SET_USER_PERMISSION",
  },
  Object {
    "type": "LOADING_START",
  },
  Object {
    "payload": Array [],
    "type": "GET_USER_SETTING",
  },
  Object {
    "type": "RELOAD_TIMESHEET",
  },
  Object {
    "payload": Object {
      "departmentCode": "",
      "departmentName": "Department Name",
      "employeeCode": "CODE",
      "employeeName": "Employee Name",
      "employeePhotoUrl": "photoUrl",
      "id": "ID",
      "managerName": "Manager Name",
      "title": "Title",
    },
    "type": "COMMONS/MODULES/PROXY_EMP_INFO/SET",
  },
  Object {
    "type": "LOADING_END",
  },
]
`;

exports[`onStampSuccess() should do. 1`] = `Array []`;

exports[`switchProxyEmployee() should do. 1`] = `
Array [
  Object {
    "type": "LOADING_START",
  },
  Object {
    "type": "RELOAD_TIMESHEET",
  },
  Object {
    "payload": Object {
      "id": "employeeId",
    },
    "type": "COMMONS/MODULES/PROXY_EMP_INFO/SET",
  },
  Object {
    "type": "LOADING_END",
  },
]
`;

exports[`switchProxyEmployee() should not do if UseCases().reloadTimesheet throw Error. 1`] = `
Array [
  Object {
    "type": "LOADING_START",
  },
  Object {
    "type": "RELOAD_TIMESHEET",
  },
  Object {
    "type": "LOADING_END",
  },
  Object {
    "payload": [Error: Error - The timetable could not be obtained. There is a possibility that the target employee does not exist on the start date of this month, or the authority necessary to switch to the target employee is insufficient.],
    "type": "CATCH_BUSINESS_ERROR",
  },
]
`;

exports[`switchProxyEmployee() should not do if error is occurred. 1`] = `
Array [
  Object {
    "type": "LOADING_START",
  },
  Object {
    "type": "RELOAD_TIMESHEET",
  },
  Object {
    "type": "LOADING_END",
  },
  Object {
    "payload": [Error: Error - The timetable could not be obtained. There is a possibility that the target employee does not exist on the start date of this month, or the authority necessary to switch to the target employee is insufficient.],
    "type": "CATCH_BUSINESS_ERROR",
  },
]
`;

exports[`switchProxyEmployee() should return if isMigratedSummary is false. 1`] = `Array []`;
